### ISSUE TITLE
twisted.web.http.Request.__repr__ is untested

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6120]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6120|
|Type|defect|
|Created|2012-10-16 13:10:37Z|
|Branch|https://github.com/twisted/twisted/tree/test-http-request-repr-6120|

It should be tested (found while porting http.py to Python 3).

Attachments:

* [3453.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7ae/7ae2243b6746416b495e16b5d0a2ed514944ae08/8e895e77052bb53054aaa33c451fcadf20441652.patch) (902 bytes) - added by moijes12 on 2013-01-20 10:10:30Z - 
* [6120.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7ae/7ae2243b6746416b495e16b5d0a2ed514944ae08/5e336e2a8c732dd3b1c3f40754c72365053a3149.patch) (1839 bytes) - added by moijes12 on 2013-01-31 18:44:37Z - 
* [6120.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7ae/7ae2243b6746416b495e16b5d0a2ed514944ae08/7cdbf1191e08c883fbd2a465d0030a5e64d18b6c.patch) (1844 bytes) - added by moijes12 on 2013-01-31 18:49:48Z - Please ignore the previous patch but the comments remain the same.
* [6120.3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7ae/7ae2243b6746416b495e16b5d0a2ed514944ae08/dfde316cab350179d7612964d24e953f5c3c84b7.patch) (1952 bytes) - added by moijes12 on 2013-02-01 05:38:31Z - 
* [6120.4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7ae/7ae2243b6746416b495e16b5d0a2ed514944ae08/2d6d3df0d35fac4e45e56f3ab2fb4f4212985a7e.patch) (1952 bytes) - added by moijes12 on 2013-02-01 05:41:49Z - Pardon me for putting y patches. But I've tried to put the best one in.
* [6120.5.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7ae/7ae2243b6746416b495e16b5d0a2ed514944ae08/f8f51765cd4f526fa43ec06d7d1401a4916572b9.patch) (1912 bytes) - added by moijes12 on 2013-03-25 18:26:46Z - Same as 6120.patch.4 but now in test_repr, a string literal is used for comparison.

<details><summary>Searchable metadata</summary>

```
trac-id__6120 6120
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_test_http_request_repr_6120 branches/test-http-request-repr-6120
branch_author__rwall rwall
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1350393037000000 1350393037000000
changetime__1365260835000000 1365260835000000
version__None None
owner__ 
cc__jknight cc__moijes12
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_3"></a>|@wallrj set owner to @moijes12|
|-|-|

# Code Review

Thanks for the patch moijes12.

It looks mostly fine. Certainly better than some of the existing
__repr__ tests I found.

But there are a few things you need to do before this can be merged...

 1. Add a News file (see [ReviewProcess#Newsfiles](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
 2. Remove all trailing whitespace from the patch (see
    http://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html#auto5)
 3. Remove "Test that " from the beginning of the test docstring...its
    superfluous (see https://plus.google.com/115348217455779620753/posts/YA3ThKWhSAj)
 4. I suppose that now this is tested, it becomes an official part of
    the http.Request API...so I think you should add a short epytext
    docstring to http.Request.__repr__ documenting the expected repr
    format.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="moijes12's avatar" src="https://avatars.githubusercontent.com/u/4566851?s=50" width="50" height="50">](https://github.com/moijes12)<a name="note_4"></a>|@moijes12 removed owner|
|-|-|

 But there are a few things you need to do before this can be merged...

  *  Add a News file (see [ReviewProcess#Newsfiles](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles)) - done
  *  Remove all trailing whitespace from the patch (see  http://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html#auto5) - I am terrible at this. I've tried but I couldn't find a way to do this.
  *  Remove "Test that " from the beginning of the test docstring...its superfluous (see  https://plus.google.com/115348217455779620753/posts/YA3ThKWhSAj) - Done.
  *  I suppose that now this is tested, it becomes an official part of the http.Request API...so I think you should add a short epytext docstring to http.Request.repr documenting the expected repr format. - Done

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra commented|
|-|-|

Why not assert with the same repr string formatting used in the actual implementation: `'<%s %s %s>'% (client.method, client.uri, client.clientproto)`

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_6"></a>|@tomprince set owner to @moijes12|
|-|-|

Thanks for your persistence in working on this:
  1. While ``__repr__`` wants to be tested, and is part of the interface (it has to be, since every python object implements it), we don't want people depending on its contents. (It is mostly there to add debugging and/or logging). Thus, we shouldn't document the format in `__repr__`s docstring (doing so in the test docstring is good).
  1. The test should use a string literal, to compare the repr against, since we have enough information to do that (c.f. point 3 [ticket:5544#comment:18 here]).
Please resubmit after addressing the above points.

**Comment by twisted-trac** [comment]
|[<img alt="moijes12's avatar" src="https://avatars.githubusercontent.com/u/4566851?s=50" width="50" height="50">](https://github.com/moijes12)<a name="note_7"></a>|@moijes12 commented|
|-|-|

Hi

1. Richard ( rwall ), in his comment, has asked for the format to be documented but tom.prince says it shouldn't. I am confused. What should I do?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

> While repr wants to be tested, and is part of the interface (it has to be, since every python object implements it), we don't want people depending on its contents. (It is mostly there to add debugging and/or logging). Thus, we shouldn't document the format in __repr__s docstring (doing so in the test docstring is good).

I disagree.  And not documenting something is a strategy with a long history of total failure as a strategy for getting people not to depend on it, anyway.

Please do say something in the docstring of `__repr__` about what its contents represent.  You don't have to be explicit about every single tiny detail, but mentioning what information about the `Request` is going to presented would be nice.

**Comment by twisted-trac** [comment]
|[<img alt="moijes12's avatar" src="https://avatars.githubusercontent.com/u/4566851?s=50" width="50" height="50">](https://github.com/moijes12)|@moijes12 removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_11"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

reviewing...

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_12"></a>|@wallrj commented|
|-|-|

(In [37802]) Branching to 'test-http-request-repr-6120'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_13"></a>|@wallrj commented|
|-|-|

(In [37803]) apply 6120.5.patch from moijes12. refs [#6120](https://github.com/twisted/twisted/issues/6120)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_14"></a>|@wallrj removed owner<br>@wallrj set status to `new`|
|-|-|

Code Review:

Thanks again moijes12 and sorry for prolonging such a simple ticket.

The latest patch looked fine to me, but when I applied it to a branch and
pointed buildbot at it I noticed a Python3 test failure.

At the moment I'm not sure what the correct remedy is so I'll ask
for some feedback from exarkun who knows more about the Python3
porting effort.

 1. [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/test-http-request-repr-6120)
    1. There's a Python3 build failure
       ```

       https://buildbot.twistedmatrix.com/builders/python-3.3-tests/builds/649/steps/shell/logs/stdio

       FAIL: test_repr (twisted.web.test.test_http.RequestTests)
       test_repr
       ----------------------------------------------------------------------
       Traceback (most recent call last):
         File "/var/lib/buildbot/buildarea/python-3_3-tests/Twisted/twisted/internet/defer.py", line 137, in maybeDeferred
           result = f(*args, **kw)
         File "/var/lib/buildbot/buildarea/python-3_3-tests/Twisted/twisted/internet/_utilspy3.py", line 41, in runWithWarningsSuppressed
           reraise(exc_info[1], exc_info[2])
         File "/var/lib/buildbot/buildarea/python-3_3-tests/Twisted/twisted/python/compat.py", line 295, in reraise
           raise exception.with_traceback(traceback)
         File "/var/lib/buildbot/buildarea/python-3_3-tests/Twisted/twisted/internet/_utilspy3.py", line 37, in runWithWarningsSuppressed
           result = f(*a, **kw)
         File "/var/lib/buildbot/buildarea/python-3_3-tests/Twisted/twisted/web/test/test_http.py", line 1665, in test_repr
           self.assertEqual(repr(request), '<GET /foo/bar HTTP/1.0>')
         File "/var/lib/buildbot/buildarea/python-3_3-tests/Twisted/twisted/trial/_synctest.py", line 356, in assertEqual
           % (msg, pformat(first), pformat(second)))
       twisted.trial.unittest.FailTest: not equal:
       a = "<b'GET' b'/foo/bar' b'HTTP/1.0'>"
       b = '<GET /foo/bar HTTP/1.0>'

       ```

       There is a discussion about whether repr should return unicode
       or bytes here: http://stackoverflow.com/questions/3627793/best-output-type-and-encoding-practices-for-repr-functions

 2. radix (on #twisted-dev) also suggested that we raise another ticket to make the
    Request.__repr__ more informative.
    ```

    19:24 < radix> rwall: hmm
    19:25 < radix> rwall: can you please include the class name and instance ID in that repr as well
    19:25 < radix> I really hate it when classes __repr__ don't include the basic info
    19:26 < radix> if you want me to put a comment on the ticket I can do that
    19:26 < radix> oh, you're reviewing, not authoring.
    19:26 < rwall> radix: Yes. That sounds like a good idea. Is it ok to change the repr without warning?
    19:27 < radix> I see. and it's not even changing the contents
    19:27 < rwall> Yep, just adding a missing test.
    19:27 < radix> rwall: yes, as tomprince said in an earlier comment on the ticket, we don't want users relying on the content of the __repr__
    19:27 < radix> but I see the content isn't even being changed in this branch, so it's probably not the appropriate branch to change it in.
    19:29 < rwall> radix: Ok, but I'll raise another ticket. What's the ideal repr? What should be the standard?
    19:30 < rwall> radix: and without wanting to dwell too long on such a small ticket...do you think __repr__ should have full docstring?
    19:32 < radix> huh, standards, that's an idea. we don't have any
    19:32 < radix> rwall: I don't think __repr__'s docstring needs to document the format absolutely. but I don't really think it matters either way.
    19:33 < rwall> ok
    19:34 < radix> Personally, off the cuff, I would suggest something like <ClassName at 0xffffffff ...> or something, where "..." could be "GET /" or                maybe even "method=GET path=/" etc

    ```

    Here are some related discussions on Stackoverflow
    * http://stackoverflow.com/questions/452300/python-object-repr-self-should-be-an-expression

I'll leave the ticket in review for now.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_15"></a>|@Julian commented|
|-|-|

Agreed on adding the class name.

As for the repr on Py3, I'd suggest just doing the string formatting in the test as well, i.e. `self.assertEqual(repr(request), '<%s %s %s>' % (method, uri, proto))

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_16"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

Some further discussion on #twisted-dev. There was reluctant agreement to allow use of string template in the test.

```
20:34 < rwall> radix, exarkun: That new web.Request.repr test fails on Python3...a unicode vs bytes issue. If you have any suggestions can you 
               comment on #6120.
20:53 < tos9> rwall: probably just do the string formatting in the test too IMHO (I'll put that on the ticket)
20:59 < rwall> tos9: Thanks I'll try that...once I've installed Python3
21:00 < tos9> heh.
21:01 -!- zooko [~user@...] has quit [Ping timeout: 245 seconds]
21:07 -!- trenton42 [~trenton@...] has quit [Ping timeout: 252 seconds]
21:26 -!- z3p [~Adium@...] has quit [Ping timeout: 252 seconds]
21:35 < rwall> tos9: Yeah that seems like the right thing to do...and that's how the test was originally implemented...but in an earlier review he'd 
               been asked to use a string literal in the test. :)
21:38 < tos9> rwall: Ah, I see :). Well, yeah think it's reasonable to go back to string formatted.
21:40 -!- hynek [~hynek@pdpc/supporter/active/hynek] has joined #twisted-dev
21:41 < tomprince> tos9 rwall: That actually looks like the test is catching a bug in python3 support.
21:41 < tomprince> Because we don't want b'' in the repr.
21:41 < tomprince> Which is exactly why having an explicit string is better in the tests.
21:44 < rwall> tomprince: let me think about that for a moment....
21:45 < glyph> tomprince: "explicit string" sounds like a contradiction :)
21:46 < glyph> tomprince: do you mean like, nativeString?
21:49 < rwall> isn't it useful to see the value type in the repr...in fact shouldn't we be using %r so that its obvious if you've assigned a unicode 
               URI
21:50 < rwall> Python3 >>> '<Request method=%r>' % (u'GET', )
21:50 < rwall> "<Request method='GET'>"
21:50 < rwall> Python2 >>> '<Request method=%r>' % (u'GET', )

21:51 < exarkun> literal string
21:52 < exarkun> rwall: The way you find out of if you totally broke your program by passing unicode where bytes is required /should not/ be by 
                 looking at the repr of the object and scanning for u in the right place
21:53 < exarkun> I think actually "<Request method=GET>" would be better than either of those strings
21:54 < rwall> exarkun: Yeah I know, but it's still useful to see the type isn't it?
21:54 < tomprince> Why?
21:56 < exarkun> rwall: I don't know.
21:56 < exarkun> Perhaps the fact that the type is /wrong/ is what leads me to think it probably doesn't need to be reflected in the repr string
21:56 < exarkun> (GET is neither bytes nor text, it is a value in the "HTTP request method" enumeration type)
21:58 < rwall> Ideally it should be one of t.p.constant?
21:58 < exarkun> yes
22:00 < rwall> If it was, then the Request.__repr__ string format would be better as %r...I think.
22:01 < rwall> tomprince, exarkun: anyway, does this reveal a bug in Python3 support. I can't see it.
22:01  * exarkun shrugs
22:02 < exarkun> an extra b in the repr isn't the end of the world, certainly

22:02 < glyph> rwall: does a test fail? :)

22:02 < exarkun> I think it's better if it's not there, but I don't _really_ care

22:05 < rwall> glyph: It does with the current test implementation, but its simple to fix.

```

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_17"></a>|@wallrj commented|
|-|-|

(In [37804]) revert to string template to make tests pass in python3. refs [#6120](https://github.com/twisted/twisted/issues/6120)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_18"></a>|@wallrj commented|
|-|-|

(In [37805]) fix some epydoc links. refs [#6120](https://github.com/twisted/twisted/issues/6120)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_19"></a>|@wallrj commented|
|-|-|

 * I changed the new test to use a string template again (as originally suggested by thijs and then again by julian.
 * Changed a couple of C{Request} to L{Request} / L{http.Request} so that the API docs have correct links.

I'll merge it if the [Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/test-http-request-repr-6120) look ok.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_20"></a>|@wallrj commented|
|-|-|

I raised [#6391](https://github.com/twisted/twisted/issues/6391) for improving the http.Request.__repr__ format.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj removed owner<br>@wallrj set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_22"></a>|@wallrj commented|
|-|-|

(In [37841]) ensure all repr attributes are unicode strings for consistent output in python 2 and python3. refs [#6120](https://github.com/twisted/twisted/issues/6120)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_23"></a>|@wallrj commented|
|-|-|

Ready for another review in log:branches/test-http-request-repr-6120

After further discussion on #twisted-dev

 * Use t.p.compat.nativeString to ensure that the values are prefixed
   with b'' in python3.
 * Which allowed me to change the test back to use a literal string
 * Added another test to demonstrate the default Request attribute
   values some of which are byte strings and some of which are native
   strings.

 * [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/test-http-request-repr-6120)


```
16:32 < tomprince> rwall: Changing tests to make them pass makes me sad.
16:33 < tomprince> If the test is *wrong* that's one thing, but I dont think that is the case here.
16:33 < rwall> I thought there was some agreement yesterday that a string template would be ok.
16:34 < rwall> tomprince: can you suggest an alternative.
16:34 < rwall> separate python2 and python3 tests?
16:34 < tomprince> Fix the code to make the test pass?
16:35 < tomprince> Its a minor point, but I think the test *is* correct.
16:35 < rwall> alright let me take another look.
16:36  * rwall searches for other repr tests
16:38 < tomprince> point 3 at https://twistedmatrix.com/trac/ticket/5544#comment:18 talks about it.
16:45 < rwall> yep, you mentioned it yesterday. But the difference is byte strings....
16:46 < rwall> take https://twistedmatrix.com/trac/browser/trunk/twisted/web/test/test_http_headers.py?annotate=blame&rev=37416#L199 for another
               recent example of using string templates in repr tests.
16:46 < tomprince> So, fix the implementation.

17:18 < kenaan_> rwall r37841 branches/test-http-request-repr-6120/twisted/web M(http.py test_http.py): ensure all repr attributes are unicode
                 strings for consistent output in python 2 and python3. refs #6120 ...
17:19 < rwall> tomprince: ^^ How about that? It looks over complicated to me.
17:19 < itamar> rwall: you don't want it to be unicode
17:19 < itamar> you want it to be a native string
17:19 < itamar> i.e. str, i.e. bytes on python 2, unicode on python 3
17:20 < itamar> luckily there's a handy function for that, twisted.python.compat.nativeString
17:20 < exarkun> also, separately, you don't ever want bytes % unicode
17:21 < tomprince> The second assertion is reasonable, but should be a seperate test.
17:21 < rwall> This is just for consistent output in __repr__ and the http.Request defaults are not all byte strings...
17:21  * rwall looks at twisted.python.compat.nativeString
17:21 < itamar> __repr__ is supposed to return a str in python 2, i.e. bytes
17:22 < exarkun> if you return unicode, Python will implicitly encode it for you (and fail if it contains non-defaultencoding encodeable characters)
17:22 < exarkun> ideally we wouldn't ever trigger implicit encoding or decoding

17:27 < kenaan_> rwall r37842 M(branches/test-http-request-repr-6120/twisted/web/http.py): use nativeString instead ...
17:30 < kenaan_> rwall r37843 M(branches/test-http-request-repr-6120/twisted/web/test/test_http.py): a separate test for the Request default values
                 ...
17:31 < rwall> tomprince, exarkun, itamar: ok thanks. I've switched to nativeString and separated the tests. I'll put that ticket back up for review.
17:32 < tomprince> You could get away with doing the conversion inline, rather than looping and doing getattr, if you wanted. (If that was part of
                   the over-complication you were commenting about.
17:33 < rwall> yeah I guess that would be clearer now.
```


-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

Thanks.  The discussion here has been incorporated into the tests added as part of [#6391](https://github.com/twisted/twisted/issues/6391).  That ticket expands the contents of the repr string and adds the necessary tests, so I'm just going to apply those changes rather than these.  Thanks again for working to figure out what to do here.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun set status to `closed`|
|-|-|

(In [37977]) Merge http-request-repr-6391

Author: introom, exarkun
Reviewer: exarkun, tom.prince
Fixes: [#6391](https://github.com/twisted/twisted/issues/6391), [#6120](https://github.com/twisted/twisted/issues/6120)

Expand the output of `twisted.web.http.Request.__repr__` to include a bit
more information.  Also add unit tests for it.
